detector/etcd/detect.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	go func() {
+		defer close(masterDetector.done)
+		<-masterDetector.client.Shutdown()
+	}()
+
+	log.V(2).Infoln("Created new detector to watch", params.hosts, params.path)
+	return masterDetector, nil
+}
+
+// call to Detect will add a changed listener. Detect may be called more than
+// once, and each time the spec'd listener will be added to the list of those receiving notifications.
+func (md *MasterDetector) Detect(f detector.MasterChanged) (err error) {
+	if f == nil {
+		// only ever install, at most, one ignoreChanged listener. multiple instances of it
+		// just consume resources and generate misleading log messages.
+		if !atomic.CompareAndSwapInt32(&md.ignoreInstalled, 0, 1) {
+			log.V(3).Infoln("ignoreChanged listener already installed")
+			return
+		}
+
+		f = ignoreChanged
+	}
+
+	log.V(3).Infoln("spawning detect()")
+	go md.detect(f)
+	return nil
+}
+
+func (md *MasterDetector) detect(f detector.MasterChanged) {
+	log.V(3).Infoln("detecting children at", md.masterDetectPath)
+
+	for {
+		isDetectorClosed := false
+		select {
+		case <-md.Done():
+			isDetectorClosed = true
+			break
+		default:
+		}
+
+		if isDetectorClosed {
+			break
+		}
+
+		started := time.Now()
+		pair, err := md.client.Get(md.masterDetectPath)
+		if err != nil {
+			log.V(1).Infoln("Get the current master info with error, master lost; due to:", err.Error())
+			md.notifyMasterChanged(nil, f)
+			md.detectedCycle(started)
+			continue
+		}
+
+		md.notifyMasterChanged(pair, f)
+
+		go md.client.Watch(md.masterDetectPath, pair.lastIndex)
+		isWatchFailed := false
+
+		for {
+			started := time.Now()
+			select {
+			case pairNode := <-md.client.MasterChangedChan():
+				md.notifyMasterChanged(pairNode, f)
+			case err := <-md.client.MasterDetectErrChan():
+				log.V(1).Infoln("Watch ended with error, master lost; error was:", err)
+				isWatchFailed = true
+				md.notifyMasterChanged(nil, f)
+				break
+			}
+			if isWatchFailed {
+				break
+			}
+			md.detectedCycle(started)
+		}
+	}
+}
+
+// Cancle closes the detector
+func (md *MasterDetector) Cancel() {
+	md.cancel()
+}
+
+// Done returns a chan that, when closed, indicates termination of the detector
+func (md *MasterDetector) Done() <-chan struct{} {
+	return md.done
+}
+
+// MinCyclePeriod is a functional option that determines the highest frequency of master change notifications
+func MinCyclePeriod(d time.Duration) detector.Option {
+	return func(di interface{}) detector.Option {
+		md := di.(*MasterDetector)
+		old := md.minDetectorCyclePeriod
+		md.minDetectorCyclePeriod = d
+		return MinCyclePeriod(old)
+	}
+}
+
+func UsedEtcdClient(usedClient ETCDInterface) detector.Option {
+	return func(di interface{}) detector.Option {
+		md := di.(*MasterDetector)
+		old := md.client
+		md.client = usedClient
+		md.cancel = usedClient.Stop
+		return UsedEtcdClient(old)
+	}
+}
+
+// detectedCycle defines the master detected cycle
+func (md *MasterDetector) detectedCycle(started time.Time) {
```

Comment:
nit: detectedCycle .. might be better named `maybeWaitAfterDetected` or something that describes what it actually does.